pkg/metrics/server.go

```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
 	}
 	degradedState.Set(0)
 }
+
+// Determine if the certificates have changed and need to be updated
+// Returns true if both files have changed AND neither is an empty file
+func certsChanged(origCertMD5 []byte, origKeyMD5 []byte) bool {
+	// Check if both files exist
+	certNotEmpty, err := fileExistsAndNotEmpty(tlsCRT)
+	if err != nil {
+		klog.Warningf("error checking if changed TLS crt file empty / exists: %v", err)
+		return false
+	}
+	keyNotEmpty, err := fileExistsAndNotEmpty(tlsKey)
+	if err != nil {
+		klog.Warningf("error checking if changed TLS key file empty / exists: %v", err)
+		return false
+	}
+
+	if !certNotEmpty || !keyNotEmpty {
+		// One of the files is missing despite some file event
+		klog.Infof("certificate or key is missing or empty. Don't rotate.")
+		return false
+	} else {
+		currentCertMD5 := checksumFile(tlsCRT)
+		currentKeyMD5 := checksumFile(tlsKey)
+
+		klog.Infof("checksums before: %x, %x. checksums after: %x, %x", origCertMD5, origKeyMD5, currentCertMD5, currentKeyMD5)
+		// Check if the non-empty certificate / key files have actually changed
+		if !bytes.Equal(origCertMD5, currentCertMD5) && !bytes.Equal(origKeyMD5, currentKeyMD5) {
+			klog.Infof("cert and key changed, we should restart the server.")
+			return true
+		}
+	}
+
+	klog.Infof("At least one of the cert / key files has not changed")
+	return false
+}
+
+//Compute the md5 checksum for a file.
+func checksumFile(fName string) []byte {
+	file, err := os.Open(fName)
+	if err != nil {
+		klog.Infof("failed to open file %v for checksum: %v", fName, err)
+	}
+	defer file.Close()
+
+	hash := md5.New()
```

Comment:
Could you use something "better" than MD5, as we're trying to minimizeMD5 usage in RHEL? sha256 for instance.

Reply:
I had sha256 originally. I (perhaps naively) switched to md5 for the slight performance improvement because I figured security was not a concern here. I switched it back as performance really shouldn't be a problem here, and it is the current standard.